state/state.go

```diff
@@ -131,6 +142,26 @@ func (state State) MakeBlock(
 	return block, block.MakePartSet(state.ConsensusParams.BlockGossip.BlockPartSizeBytes)
 }
 
+// MedianTime computes a median time for a given Commit (based on Timestamp field of votes messages) and the
+// corresponding validator set. The computed time is always between timestamps of
+// the votes sent by honest processes, i.e., a faulty processes can not arbitrarily increase or decrease the
+// computed value.
+func MedianTime(commit *types.Commit, validators *types.ValidatorSet) time.Time {
+
+	weightedTimes := make([]*tmtime.WeightedTime, len(commit.Precommits))
+	totalVotingPower := int64(0)
+
+	for i, vote := range commit.Precommits {
+		if vote != nil {
+			_, validator := validators.GetByIndex(vote.ValidatorIndex)
+			totalVotingPower += validator.VotingPower
+			weightedTimes[i] = tmtime.NewWeightedTime(vote.Timestamp, validator.VotingPower)
+		}
+	}
+
+	return tmtime.WeightedMedian(weightedTimes, totalVotingPower)
```

Comment:
can we just call vals.TotalVotingPower() instead of accumulating it ourselves here?